### ISSUE TITLE
cypress run doesn't return to command line

### ISSUE THREAD

**Issue by paulpruteanu** [cypress-io/cypress#815]
### Current behavior:
```bash
cypress run
```
does not return to command line, under some circumstances, irrespective of the test results.
### Desired behavior:
to return to command line, as the exit code is relevant when creating a build.

### How to reproduce:
I could not replicate it, but the sequence of events in GUI is:
VISIT url # containing a hash-bang
XHR GET 200
XHR POST 404
XHR GET 200 # removes part of the URL hash-bang
CONTAINS something
XHR POST 200
XHR POST 200 # that's when the DOM get's updated with something
XHR GET 200
NEW URL # restoring the hash-bang

It works fine in the GUI, on both Chrome and Electron, but it's stuck in CLI with default browser (Electron).

### Additional Info (images, stack traces, etc)

- Operating System: MacOS Sierra 10.12.6
- Cypress Version: 1.0.2
- Browser Version: Electron 53

**Comment by youknowriad** [comment]
We're trying to use Cypress for [WordPress's Gutenberg editor](https://github.com/WordPress/gutenberg/pull/3069) and we're experiencing the same issue. (See the travis job here https://travis-ci.org/WordPress/gutenberg/jobs/293142523)

**Comment by bahmutov** [comment]
Hmm, interesting, can we have diagnostic log - if you run cypress with `DEBUG=cypress:*` environment variable it will show a lot more

**Comment by youknowriad** [comment]
Sure

<img width="1047" alt="screen shot 2017-10-26 at 15 03 08" src="https://user-images.githubusercontent.com/272444/32057363-e73aa018-ba5e-11e7-84f0-ade48a7ca3d7.png">

**Comment by bahmutov** [comment]
@brian-mann could it be timer problem? Because what we usually see after reporting test results is this
```
  - Cypress Version: 1.0.2
  cypress:server:timers queuing timer id 25 after 30000 ms +83ms
  cypress:server:timers child received timer id 25 +79ms
  cypress:server:timers clearing timer id 25 from queue { '24': { args: [], ms: 85000, cb: [Function] }, '25': { args: [], ms: 30000, cb: [Function: bound cleanupWebsocketResources] } } +3ms
  cypress:server:timers clearing timer id 24 from queue { '24': { args: [], ms: 85000, cb: [Function] } } +2ms
  cypress:server:timers clearing timer id 24 from queue {} +0ms
  cypress:server:timers child sending timer id 9 +89ms


  (Video)

  - Started processing:   Compressing to 32 CRF
  - Finished processing:  /tmp/cypress-test-tiny/cypress/videos/d3tft.mp4 (0 seconds)
  cypress:server Should copy Circle Artifacts? undefined +927ms


  (All Done)

  cypress:server about to exit with code 0 +2ms
```

**Comment by gziolo** [comment]
> We're trying to use Cypress for WordPress's Gutenberg editor and we're experiencing the same issue. (See the travis job here https://travis-ci.org/WordPress/gutenberg/jobs/293142523)

I tried to run the same tests with Chrome using `cypress run --browser chrome` and it returns to the command line just after finishing all tests.

When I execute test with `--headed` flag I see the following list of requests in the Network tab:

![screen shot 2017-10-27 at 09 18 35](https://user-images.githubusercontent.com/699132/32092257-ecb8a794-baf7-11e7-97a8-96b40378820a.png)

I also see this error on the console:

![screen shot 2017-10-27 at 09 22 29](https://user-images.githubusercontent.com/699132/32092354-64ed304a-baf8-11e7-95c8-1c31c2bf3f1b.png)

**Comment by belohlavek** [comment]
I'm having the same issue, how did you guys solve this?

**Comment by gziolo** [comment]
Use chrome instead of electron as the selected browser 😎

**Comment by brian-mann** [comment]
I've seen an issue in TravisCI before with it not killing background processes per command. Basically those will continue to keep it open.

I'd try splitting up booting + backgrounding your webserver from running the Cypress process to see if that fixes it.

**Comment by youknowriad** [comment]
@brian-mann I just want to precise than It happens locally as well. (MacOS 10.13)

**Comment by paulpruteanu** [comment]
I found the issue on my end.
If your test looks like describe -> describe or context of a description larger than 176chars, cypress run hangs on a failed test, never returning to cli, irrespective of the browser.

**Comment by bahmutov** [comment]
Wait, can you put a code block that does this into the issue - so we can reproduce this unexpected bug

**Comment by paulpruteanu** [comment]
```js
describe(`my feature file`, () => {

    context(`that has a describe/ context/ it clause(s) which altogether contains
a combined description string length of 250 characters, when your test fails for
some reason or another`, () => {
        it(`halts, without showing the error, or returning to the CLI prompt`, () => {
            cy.visit('https://www.cypress.io/')
            cy.contains('Finally').should('not.exist')
        })
    })
})
/* `my feature filethat has a describe/ context/ it clause(s) which altogether contains
a combined description string length of 250 characters, when your test fails for
some reason or anotherhalts, without showing the error, or returning to the CLI prompt`
.length == 251. try remove 2 characters from any of the describe/ context/ it, and
the error will appear, alongside with the return to the command prompt.*/
```

**Comment by belohlavek** [comment]
For me it happens with Electron headless and 

```
describe('Log Viewer', () => {
  beforeEach(() => {
    cy.visit(Cypress.env('HOST'));
  });

  it('should open the Log Viewer', () => {
    cy.getElementByTestId('log-viewer-tab').click();
    cy.getElementByTestId('log-viewer-notice').should('be.visible');
  });
});
```

Both locally on Mac and on Travis

**Comment by paulpruteanu** [comment]
Another issue I'm facing is related to beforeunload (#796).

```html
<!-- index.html -->
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
    <script>
        window.pageRefreshConfirmation = (e) => {
            (e||window.event).returnValue = "Are you sure?";
        }
        window.addEventListener("beforeunload", window.pageRefreshConfirmation)
    </script>
</head>
<body>
    foo bar
</body>
</html>
```

```js
// cypress/integration/test.js
describe(`beforeunload issue`, () => {

    context(`electron browser fails to return to CLI on a page with beforeunload,
even when removing the listener`, () => {
        it(`checks foo`, () => {
            cy.visit('http://127.0.0.1:8080/')
                .then(() => {
                    cy.window().then(win => win.removeEventListener(
                        'beforeunload',
                        win.pageRefreshConfirmation)
                    )
                })
            cy.contains('foo').should('be.visible')
        })
        it(`checks bar`, () => {
            cy.visit('http://127.0.0.1:8080/')
            cy.contains('bar').should('be.visible')
        })
    })
})
```
when running
```unix
./node_modules/.bin/cypress run --spec cypress/integration/test.js
```
you hang up, not getting returned to command prompt:
<img width="527" alt="screen shot 2017-11-07 at 11 44 06" src="https://user-images.githubusercontent.com/8288389/32492116-0e5438d4-c3b1-11e7-8cfb-b697327c221e.png">

If I remove the event listener from index.html:
```html
<script>
    window.pageRefreshConfirmation = (e) => {
        (e||window.event).returnValue = "If you continue, any information you have entered will not be saved.";
    }
    //window.addEventListener("beforeunload", window.pageRefreshConfirmation)
</script>
```
then I am returned to the command prompt
<img width="584" alt="screen shot 2017-11-07 at 11 52 13" src="https://user-images.githubusercontent.com/8288389/32492425-25969dba-c3b2-11e7-8252-4d38e439aa2a.png">

**Comment by fsodano** [comment]
Thanks for your detailed report @paulpruteanu, it allowed me to solve this issue with a small workaround:

I simply clear the `onbeforeunload` function from the `window`. I am running it `beforeEach` although I suspect this is an overkill.

```javascript 
beforeEach(function () {
    cy.window().then(win => win.onbeforeunload = undefined);
});
```

**Comment by krishnaxv** [comment]
@brian-mann @bahmutov I am also facing this issue in GitLab CI. All the tests are running fine but the control is not returning back to CLI.

**Is this an issue with Cypress or I am doing something wrong?**

I have attached a screenshot of this issue.

![cypress ci build](https://user-images.githubusercontent.com/8594438/37343218-de5c3f06-26ec-11e8-9ad2-b038535edb67.png)

Any help or direction will be greatly appreciated. Thanks!

**Comment by mgalindo** [comment]
I am experiencing the same issue running locally in OS X

**Comment by igor-starostenko-ag** [comment]
I've had the same issue when running my tests with Electron.
In my case it was caused by a browser warning message that shows up when you're trying to close the page.
Like this one: 
![leave_site_prompt](https://user-images.githubusercontent.com/28714643/40560083-bd318b10-600d-11e8-937a-224b5eea84c5.png)
I had to add some steps to navigate to another page that doesn't trigger this warning. Then the process could exit properly.
Hope this can help.
